File: backend/payments.py
"""
Pipways — Paystack Payment Integration (Nigeria market)
Replaces Stripe entirely.

Endpoints:
  POST /payments/initiate          — start a subscription checkout
  POST /payments/verify/{ref}      — verify after Paystack redirect
  GET  /payments/plans             — list available plans with Naira pricing
  POST /payments/webhook           — Paystack webhook (charge.success events)

Environment variables required:
  PAYSTACK_SECRET_KEY   — sk_live_... or sk_test_...
  PAYSTACK_PUBLIC_KEY   — pk_live_... or pk_test_...
  APP_URL               — e.g. https://pipways.com (for callback URL)
"""

import os
import hmac
import hashlib
import json
import httpx
from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel
from typing import Optional
from datetime import datetime

from .database import database
from .security import get_current_user, get_user_id

router = APIRouter()

PAYSTACK_SECRET_KEY = os.getenv("PAYSTACK_SECRET_KEY", "")
PAYSTACK_PUBLIC_KEY = os.getenv("PAYSTACK_PUBLIC_KEY", "")
APP_URL = os.getenv("APP_URL", "https://pipways.com")
PAYSTACK_BASE = "https://api.paystack.co"


# ── Pricing in Naira ─────────────────────────────────────────────────────────
# Trading Academy is FREE for all users — no plan needed for that.
# These plans gate AI tools and pro features only.

PLANS = {
    "pro_monthly": {
        "name":        "Pro Monthly",
        "tier":        "pro",
        "amount_ngn":  20000,
        "amount_kobo": 20000 * 100,
        "interval":    "monthly",
        "description": "Unlimited AI Mentor, Chart Analysis, Performance Analytics, Signals + Telegram",
        "features": [
            "Unlimited AI Mentor sessions",
            "20 Chart Analyses per month",
            "Unlimited Performance Analytics",
            "Full trading signals + Telegram alerts",
            "Webinar recordings access",
        ],
    },
    "pro_yearly": {
        "name":        "Pro Yearly",
        "tier":        "pro",
        "amount_ngn":  200000,
        "amount_kobo": 200000 * 100,
        "interval":    "annually",
        "description": "Everything in Pro, billed yearly. Save \u20a640,000.",
        "features": [
            "Everything in Pro Monthly",
            "Save \u20a640,000 vs monthly billing",
            "Priority AI response speed",
        ],
    },
    "power_monthly": {
        "name":        "Power Trader",
        "tier":        "pro_plus",
        "amount_ngn":  35000,
        "amount_kobo": 35000 * 100,
        "interval":    "monthly",
        "description": "Everything Pro + unlimited Chart Analysis + AI Stock Terminal",
        "features": [
            "Everything in Pro",
            "Unlimited Chart Analyses",
            "AI Stock Research Terminal",
            "Earliest access to new features",
        ],
    },
}


# ── Models ────────────────────────────────────────────────────────────────────

class InitiatePaymentRequest(BaseModel):
    plan_key: str
    coupon: Optional[str] = None


# ── Helpers ───────────────────────────────────────────────────────────────────

def _headers() -> dict:
    if not PAYSTACK_SECRET_KEY:
        raise HTTPException(503, "Payment system not configured. Contact support.")
    return {
        "Authorization": f"Bearer {PAYSTACK_SECRET_KEY}",
        "Content-Type": "application/json",
    }


async def _paystack_post(path: str, payload: dict) -> dict:
    async with httpx.AsyncClient(timeout=30) as client:
        res = await client.post(f"{PAYSTACK_BASE}{path}", headers=_headers(), json=payload)
    data = res.json()
    if not data.get("status"):
        raise HTTPException(400, data.get("message", "Paystack error"))
    return data


async def _paystack_get(path: str) -> dict:
    async with httpx.AsyncClient(timeout=30) as client:
        res = await client.get(f"{PAYSTACK_BASE}{path}", headers=_headers())
    data = res.json()
    if not data.get("status"):
        raise HTTPException(400, data.get("message", "Paystack error"))
    return data


# ── Routes ────────────────────────────────────────────────────────────────────

@router.get("/config")
async def get_config():
    """Return Paystack public key for frontend Inline JS."""
    if not PAYSTACK_PUBLIC_KEY:
        raise HTTPException(503, "Payment system not configured")
    return {"public_key": PAYSTACK_PUBLIC_KEY}


@router.get("/plans")
async def get_plans():
    """
    Return all available plans with Naira pricing keyed by plan key.
    Public — no auth required (used on pricing page & subscription section).
    """
    return {
        key: {
            "name":        plan["name"],
            "tier":        plan["tier"],
            "amount_ngn":  plan["amount_ngn"],
            "amount":      plan["amount_kobo"],
            "interval":    plan["interval"],
            "description": plan["description"],
            "features":    plan["features"],
        }
        for key, plan in PLANS.items()
    }


@router.post("/initiate")
async def initiate_payment(
    body: InitiatePaymentRequest,
    current_user=Depends(get_current_user),
):
    """
    Start a Paystack checkout.
    Returns authorization_url — redirect the user there.

    Flow:
      1. POST /payments/initiate  → { authorization_url, reference }
      2. Redirect user to authorization_url
      3. Paystack redirects to APP_URL/payments/callback?reference=...
      4. POST /payments/verify/{reference}  → tier upgraded
    """
    plan = PLANS.get(body.plan_key)
    if not plan:
        raise HTTPException(400, f"Unknown plan: {body.plan_key}. Valid: {list(PLANS.keys())}")

    user_id = get_user_id(current_user)
    email = (
        current_user.get("email") if isinstance(current_user, dict)
        else getattr(current_user, "email", None)
    )
    if not email:
        raise HTTPException(400, "Could not determine user email")

    reference = f"pip_{user_id}_{plan['tier']}_{int(datetime.utcnow().timestamp())}"

    try:
        await database.execute(
            """
            INSERT INTO payment_transactions
                (user_id, reference, plan_key, tier, amount_ngn, status, created_at)
            VALUES (:uid, :ref, :plan, :tier, :amount, 'pending', NOW())
            ON CONFLICT (reference) DO NOTHING
            """,
            {"uid": user_id, "ref": reference, "plan": body.plan_key,
             "tier": plan["tier"], "amount": plan["amount_ngn"]},
        )
    except Exception as e:
        print(f"[PAYMENTS] Could not store pending transaction: {e}", flush=True)

    data = await _paystack_post(
        "/transaction/initialize",
        {
            "email":        email,
            "amount":       plan["amount_kobo"],
            "reference":    reference,
            "callback_url": f"{APP_URL}/payments/callback",
            "metadata": {
                "user_id":  user_id,
                "plan_key": body.plan_key,
                "tier":     plan["tier"],
            },
            "channels": ["card", "bank", "ussd", "mobile_money", "bank_transfer"],
        },
    )

    return {
        "authorization_url": data["data"]["authorization_url"],
        "reference":         data["data"]["reference"],
        "plan":              plan["name"],
        "amount_ngn":        plan["amount_ngn"],
    }


@router.post("/verify/{reference}")
async def verify_payment(reference: str, current_user=Depends(get_current_user)):
    """
    Verify a Paystack payment after redirect.
    Upgrades the user's subscription_tier on success.
    """
    user_id = get_user_id(current_user)
    data = await _paystack_get(f"/transaction/verify/{reference}")
    tx = data["data"]

    if tx["status"] != "success":
        raise HTTPException(400, f"Payment not successful. Status: {tx['status']}")

    meta = tx.get("metadata", {})
    plan_key = meta.get("plan_key")
    tier = meta.get("tier") or (PLANS.get(plan_key, {}).get("tier") if plan_key else None)
    if not tier:
        raise HTTPException(400, "Could not determine tier from payment metadata")

    await _upgrade_user(user_id, tier, reference)

    return {
        "success":   True,
        "tier":      tier,
        "reference": reference,
        "message":   f"Payment verified. You now have {tier} access.",
    }


@router.post("/webhook")
async def paystack_webhook(request: Request):
    """
    Paystack webhook for charge.success events.
    Add this URL in Paystack dashboard → Settings → Webhooks:
      https://pipways.com/payments/webhook
    """
    signature = request.headers.get("x-paystack-signature", "")
    body_bytes = await request.body()

    if not PAYSTACK_SECRET_KEY:
        print("[PAYMENTS] ⚠️ PAYSTACK_SECRET_KEY not set — rejecting webhook", flush=True)
        raise HTTPException(503, "Payment system not configured")

    expected = hmac.new(
        PAYSTACK_SECRET_KEY.encode("utf-8"),
        body_bytes,
        hashlib.sha512,
    ).hexdigest()
    if not hmac.compare_digest(signature, expected):
        raise HTTPException(400, "Invalid webhook signature")

    payload = json.loads(body_bytes)

    if payload.get("event") == "charge.success":
        tx_data = payload.get("data", {})
        reference = tx_data.get("reference", "")
        meta = tx_data.get("metadata", {})
        user_id = meta.get("user_id")
        plan_key = meta.get("plan_key")
        tier = meta.get("tier") or (PLANS.get(plan_key, {}).get("tier") if plan_key else None)

        if user_id and tier:
            await _upgrade_user(int(user_id), tier, reference)
            print(f"[PAYMENTS] Webhook: upgraded user {user_id} to {tier}", flush=True)

    return {"status": "ok"}


# ── Internal ──────────────────────────────────────────────────────────────────

async def _upgrade_user(user_id: int, tier: str, reference: str) -> None:
    """Upgrade user tier. Idempotent — safe to call from /verify and /webhook."""
    try:
        await database.execute(
            """UPDATE users SET subscription_tier = :tier,
               subscription_updated_at = NOW() WHERE id = :uid""",
            {"tier": tier, "uid": user_id},
        )
    except Exception:
        try:
            await database.execute(
                "UPDATE users SET subscription_tier = :tier WHERE id = :uid",
                {"tier": tier, "uid": user_id},
            )
        except Exception as e:
            print(f"[PAYMENTS] Failed to upgrade user {user_id}: {e}", flush=True)
            return

    try:
        await database.execute(
            """UPDATE payment_transactions SET status='completed', completed_at=NOW()
               WHERE reference=:ref""",
            {"ref": reference},
        )
    except Exception as e:
        print(f"[PAYMENTS] Could not update transaction status: {e}", flush=True)

    print(f"[PAYMENTS] User {user_id} upgraded to {tier} (ref: {reference})", flush=True)


async def init_payment_tables():
    """Create payment_transactions table. Call from main.py lifespan."""
    try:
        await database.execute("""
            CREATE TABLE IF NOT EXISTS payment_transactions (
                id           SERIAL PRIMARY KEY,
                user_id      INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
                reference    VARCHAR(120) UNIQUE NOT NULL,
                plan_key     VARCHAR(60) NOT NULL,
                tier         VARCHAR(30) NOT NULL,
                amount_ngn   INTEGER NOT NULL,
                status       VARCHAR(20) NOT NULL DEFAULT 'pending',
                created_at   TIMESTAMP NOT NULL DEFAULT NOW(),
                completed_at TIMESTAMP
            )
        """)
        await database.execute(
            "ALTER TABLE users ADD COLUMN IF NOT EXISTS subscription_updated_at TIMESTAMP"
        )
        print("[PAYMENTS] Payment tables ready.", flush=True)
    except Exception as e:
        print(f"[PAYMENTS] Table init warning: {e}", flush=True)

File: frontend/js/payments.js
/**
 * payments.js — Paystack Integration for Pipways
 *
 * Loads plans from /payments/plans, shows a pricing modal,
 * then uses Paystack Inline to collect payment without leaving the page.
 *
 * Dependencies (add to dashboard.html before this file):
 *   <script src="https://js.paystack.co/v1/inline.js"></script>
 *
 * Usage:
 *   PaymentsPage.showUpgradeModal()          — from any 402 error handler
 *   PaymentsPage.render(container)           — full pricing page
 */

// ── Inline helpers (avoid dependency on ui.js / store.js) ──────────────────
const _pw = {
    _toastBox: null,
    _ensureToastBox() {
        if (this._toastBox) return this._toastBox;
        let box = document.getElementById('pw-toast-container');
        if (!box) {
            box = document.createElement('div');
            box.id = 'pw-toast-container';
            box.style.cssText = 'position:fixed;bottom:2rem;right:2rem;z-index:9999;display:flex;flex-direction:column;gap:.75rem;';
            document.body.appendChild(box);
        }
        this._toastBox = box;
        return box;
    },
    toast(msg, type = 'info') {
        const colors = { success:'#10b981', error:'#ef4444', warning:'#f59e0b', info:'#3b82f6' };
        const t = document.createElement('div');
        t.style.cssText = `background:#1f2937;color:#f3f4f6;border-left:4px solid ${colors[type]||colors.info};padding:.75rem 1.25rem;border-radius:.5rem;box-shadow:0 4px 12px rgba(0,0,0,.3);min-width:260px;max-width:400px;display:flex;justify-content:space-between;align-items:center;`;
        t.innerHTML = `<span>${msg}</span><button onclick="this.parentElement.remove()" style="background:none;border:none;cursor:pointer;font-size:1.25rem;color:#9ca3af;margin-left:.75rem;">&times;</button>`;
        this._ensureToastBox().appendChild(t);
        setTimeout(() => { if (t.parentElement) t.remove(); }, 5000);
    },
    modal(content) {
        const m = document.createElement('div');
        m.className = 'pw-modal-overlay';
        m.style.cssText = 'position:fixed;top:0;left:0;width:100%;height:100%;background:rgba(0,0,0,.6);z-index:1000;display:flex;align-items:center;justify-content:center;padding:1rem;';
        m.innerHTML = `<div style="background:#1f2937;color:#f3f4f6;border-radius:1rem;max-width:600px;width:100%;max-height:90vh;overflow-y:auto;box-shadow:0 25px 50px -12px rgba(0,0,0,.4);position:relative;">
            <button onclick="this.closest('.pw-modal-overlay').remove()" style="position:absolute;top:1rem;right:1rem;background:none;border:none;font-size:1.5rem;cursor:pointer;color:#9ca3af;">&times;</button>
            <div style="padding:2rem;">${content}</div></div>`;
        m.addEventListener('click', e => { if (e.target === m) m.remove(); });
        document.body.appendChild(m);
    },
    closeModal() { document.querySelectorAll('.pw-modal-overlay').forEach(m => m.remove()); }
};

// Provide global UI/Router shims if they don't already exist
if (typeof UI === 'undefined') {
    window.UI = { showToast: (m,t) => _pw.toast(m,t), showModal: c => _pw.modal(c), closeModal: () => _pw.closeModal() };
}
if (typeof Router === 'undefined') {
    window.Router = { navigate(path) { window.location.href = path; } };
}

const PaymentsPage = {

    _plans: null,
    _publicKey: null,

    // ── Hardcoded fallback plans (shown if backend is unreachable) ────────
    _fallbackPlans: {
        pro_monthly: {
            name: 'Pro Monthly', tier: 'pro', amount_ngn: 5000, amount: 500000,
            interval: 'monthly', description: 'Unlimited AI Mentor, Chart Analysis, Performance Analytics, Signals + Telegram',
            features: ['Unlimited AI Mentor sessions', '20 Chart Analyses per month', 'Unlimited Performance Analytics', 'Full trading signals + Telegram alerts', 'Webinar recordings access']
        },
        pro_yearly: {
            name: 'Pro Yearly', tier: 'pro', amount_ngn: 45000, amount: 4500000,
            interval: 'annually', description: 'Everything in Pro, billed yearly. Save N15,000.',
            features: ['Everything in Pro Monthly', 'Save N15,000 vs monthly billing', 'Priority AI response speed']
        },
        power_monthly: {
            name: 'Power Trader', tier: 'pro_plus', amount_ngn: 12000, amount: 1200000,
            interval: 'monthly', description: 'Everything Pro + unlimited Chart Analysis + AI Stock Terminal',
            features: ['Everything in Pro', 'Unlimited Chart Analyses', 'AI Stock Research Terminal', 'Earliest access to new features']
        }
    },

    // ── Load config from backend (with timeout) ────────────────────────────
    async _loadConfig() {
        if (this._plans && this._publicKey) return;

        const timeout = (ms) => new Promise((_, reject) =>
            setTimeout(() => reject(new Error('timeout')), ms));

        try {
            const config = await Promise.race([
                API.request('/payments/config'),
                timeout(8000)
            ]);
            this._publicKey = config.public_key;
        } catch (e) {
            console.warn('[Payments] Config unavailable:', e.message || e);
        }

        try {
            const full = await Promise.race([
                API.request('/payments/plans'),
                timeout(8000)
            ]);
            if (full && typeof full === 'object' && Object.keys(full).length > 0) {
                this._plans = full;
            }
        } catch (e) {
            console.warn('[Payments] Plans unavailable, using fallback:', e.message || e);
        }

        // Always fall back to hardcoded plans if backend didn't return any
        if (!this._plans) {
            this._plans = this._fallbackPlans;
        }
    },

    // ── Full pricing page ───────────────────────────────────────────────────
    async render(container) {
        container.innerHTML = `
            <div class="page-header">
                <h1>Upgrade Your Account</h1>
                <p style="color:#6b7280;">Choose a plan to unlock AI tools, signals, and more</p>
            </div>
            <div id="plans-grid" style="display:grid;grid-template-columns:repeat(auto-fit,minmax(280px,1fr));gap:1.5rem;margin-top:2rem;">
                <div style="text-align:center;padding:3rem;color:#9ca3af;">
                    <i class="fas fa-spinner fa-spin text-2xl"></i>
                    <p style="margin-top:1rem;">Loading plans…</p>
                </div>
            </div>
        `;
        try {
            await this._loadConfig();
        } catch (e) {
            console.error('[Payments] render failed:', e);
        }
        const grid = document.getElementById('plans-grid');
        if (grid) this._renderPlanCards(grid);
    },

    _renderPlanCards(container) {
        if (!this._plans) {
            container.innerHTML = '<p style="color:#ef4444;">Failed to load plans. Please try again.</p>';
            return;
        }

        const user = JSON.parse(localStorage.getItem('pipways_user') || 'null');
        const currentTier = user?.subscription_tier || 'free';

        const planOrder = ['pro_monthly', 'pro_yearly', 'power_monthly'];
        const badges = {
            'pro_yearly': { text: 'Best Value', color: '#10b981' },
            'pro_monthly': { text: 'Most Popular', color: '#3b82f6' },
        };

        container.innerHTML = planOrder.map(key => {
            const plan = this._plans[key];
            if (!plan) return '';
            const isActive = currentTier === plan.tier;
            const badge = badges[key];

            return `
            <div style="background:#111827;border-radius:1rem;padding:2rem;border:2px solid ${isActive ? '#3b82f6' : '#1f2937'};
                        position:relative;box-shadow:0 1px 3px rgba(0,0,0,.3);transition:box-shadow .2s, border-color .2s;"
                 onmouseover="this.style.boxShadow='0 10px 25px rgba(0,0,0,.4)';this.style.borderColor='#374151'"
                 onmouseout="this.style.boxShadow='0 1px 3px rgba(0,0,0,.3)';this.style.borderColor='${isActive ? '#3b82f6' : '#1f2937'}'">

                ${badge ? `<div style="position:absolute;top:-12px;left:50%;transform:translateX(-50%);
                    background:${badge.color};color:white;font-size:.75rem;font-weight:700;
                    padding:.25rem .75rem;border-radius:99px;">${badge.text}</div>` : ''}

                ${isActive ? `<div style="position:absolute;top:-12px;right:1rem;
                    background:#10b981;color:white;font-size:.75rem;font-weight:700;
                    padding:.25rem .75rem;border-radius:99px;">Current Plan</div>` : ''}

                <h3 style="font-size:1.25rem;font-weight:700;color:#f9fafb;margin-bottom:.5rem;">
                    ${plan.name}
                </h3>
                <p style="color:#9ca3af;font-size:.875rem;margin-bottom:1.5rem;">${plan.description}</p>

                <div style="margin-bottom:1.5rem;">
                    <span style="font-size:2.25rem;font-weight:800;color:#f9fafb;">
                        ₦${plan.amount_ngn.toLocaleString()}
                    </span>
                    <span style="color:#9ca3af;font-size:.875rem;">
                        /${plan.interval === 'annually' ? 'year' : 'month'}
                    </span>
                </div>

                <ul style="list-style:none;padding:0;margin-bottom:1.5rem;">
                    ${(plan.features || []).map(f => `
                        <li style="display:flex;align-items:center;gap:.5rem;padding:.375rem 0;
                                   font-size:.875rem;color:#d1d5db;border-bottom:1px solid #1f2937;">
                            <i class="fas fa-check" style="color:#10b981;flex-shrink:0;"></i>
                            ${f}
                        </li>
                    `).join('')}
                </ul>

                ${isActive
                    ? `<button disabled style="width:100%;padding:.75rem;border-radius:.5rem;
                          background:#1f2937;color:#9ca3af;border:none;font-weight:600;cursor:default;">
                          Active Plan
                       </button>`
                    : `<button onclick="PaymentsPage.startPayment('${key}')"
                          style="width:100%;padding:.75rem;border-radius:.5rem;
                                 background:#3b82f6;color:white;border:none;font-weight:700;
                                 cursor:pointer;font-size:.9rem;transition:background .2s;"
                          onmouseover="this.style.background='#2563eb'"
                          onmouseout="this.style.background='#3b82f6'">
                          Upgrade to ${plan.tier.charAt(0).toUpperCase() + plan.tier.slice(1)}
                       </button>`
                }
            </div>`;
        }).join('');
    },

    // ── Quick upgrade modal (shown on 402 errors) ───────────────────────────
    async showUpgradeModal(featureName = 'this feature') {
        await this._loadConfig();

        const user = JSON.parse(localStorage.getItem('pipways_user') || 'null');
        const currentTier = user?.subscription_tier || 'free';

        // Suggest pro_monthly as default upgrade
        const suggestedKey = currentTier === 'pro' ? 'power_monthly' : 'pro_monthly';
        const plan = this._plans?.[suggestedKey];

        UI.showModal(`
            <div style="text-align:center;padding:1rem 0 1.5rem;">
                <div style="font-size:2.5rem;margin-bottom:1rem;">🔒</div>
                <h2 style="font-size:1.4rem;font-weight:700;color:#111827;margin-bottom:.5rem;">
                    Upgrade to unlock ${featureName}
                </h2>
                <p style="color:#6b7280;font-size:.9rem;margin-bottom:1.5rem;">
                    You've reached your free tier limit.
                </p>

                ${plan ? `
                <div style="background:#f8fafc;border:1px solid #e2e8f0;border-radius:.75rem;
                            padding:1.25rem;margin-bottom:1.5rem;text-align:left;">
                    <div style="font-weight:700;color:#111827;">${plan.name}</div>
                    <div style="font-size:1.5rem;font-weight:800;color:#3b82f6;margin:.25rem 0;">
                        ₦${plan.amount_ngn?.toLocaleString()}<span style="font-size:.875rem;color:#9ca3af;">/${plan.interval === 'annually' ? 'year' : 'month'}</span>
                    </div>
                    <ul style="list-style:none;padding:0;margin:.75rem 0 0;">
                        ${(plan.features || []).slice(0, 3).map(f => `
                            <li style="font-size:.8rem;color:#374151;padding:.2rem 0;">
                                <i class="fas fa-check" style="color:#10b981;margin-right:.4rem;"></i>${f}
                            </li>
                        `).join('')}
                    </ul>
                </div>

                <button onclick="UI.closeModal(); PaymentsPage.startPayment('${suggestedKey}')"
                    style="width:100%;padding:.85rem;border-radius:.5rem;background:#3b82f6;
                           color:white;border:none;font-weight:700;cursor:pointer;font-size:1rem;">
                    Upgrade Now — ₦${plan.amount_ngn?.toLocaleString()}
                </button>
                <button onclick="UI.closeModal()"
                    style="width:100%;margin-top:.75rem;padding:.7rem;border-radius:.5rem;
                           background:none;color:#9ca3af;border:1px solid #e5e7eb;cursor:pointer;">
                    Maybe Later
                </button>
                ` : `
                <button onclick="UI.closeModal(); Router.navigate('/pricing')"
                    style="width:100%;padding:.85rem;border-radius:.5rem;background:#3b82f6;
                           color:white;border:none;font-weight:700;cursor:pointer;">
                    View Plans
                </button>`}
            </div>
        `);
    },

    // ── Paystack Inline popup ───────────────────────────────────────────────
    async startPayment(planKey) {
        await this._loadConfig();

        const user = JSON.parse(localStorage.getItem('pipways_user') || 'null');
        if (!user) {
            UI.showToast('Please log in to upgrade', 'warning');
            return;
        }

        if (!this._publicKey) {
            UI.showToast('Payment system unavailable — contact support', 'error');
            return;
        }

        const plan = this._plans?.[planKey];
        if (!plan) {
            UI.showToast('Plan not found', 'error');
            return;
        }

        // Get a reference from backend (backend stores metadata)
        let reference;
        try {
            const init = await API.request('/payments/initiate', {
                method: 'POST',
                body: JSON.stringify({ plan_key: planKey }),
            });
            reference = init.reference;
        } catch (e) {
            UI.showToast('Could not start payment: ' + e.message, 'error');
            return;
        }

        // Paystack Inline popup
        const handler = PaystackPop.setup({
            key:       this._publicKey,
            email:     user.email,
            amount:    plan.amount,          // kobo
            currency:  'NGN',
            ref:       reference,
            metadata: {
                user_id:  user.id,
                plan_key: planKey,
            },
            onClose() {
                UI.showToast('Payment window closed', 'info');
            },
            callback: async (response) => {
                // Verify on backend
                try {
                    const result = await API.request(`/payments/verify/${response.reference}`);
                    UI.showToast(result.message || 'Payment successful!', 'success');

                    // Update local user object with new tier
                    const updated = { ...user, subscription_tier: result.tier };
                    localStorage.setItem('pipways_user', JSON.stringify(updated));

                    // Redirect to dashboard
                    setTimeout(() => Router.navigate('/dashboard'), 1500);
                } catch (err) {
                    UI.showToast('Payment received but verification failed — contact support', 'warning');
                }
            },
        });

        handler.openIframe();
    },
};

window.PaymentsPage = PaymentsPage;

// ── REMOVED: 402 handler monkey-patch ────────────────────────────────────────
// This monkey-patch of API.request has been removed to avoid conflicts with the
// canonical 402 handler in /js/api.js. The API.request method now handles all
// 402 responses directly and calls PaymentsPage.showUpgradeModal when needed.
// See: /js/api.js lines ~50-66 for the primary 402 handler
